### PR TITLE
bpo-38387: Formally document PyDoc_STRVAR and PyDoc_STR macros

### PR DESCRIPTION
**Pull request by bsolomon1124**
Adds a short description of `PyDoc_STRVAR` and `PyDoc_STR` to "Useful macros" section of C-API docs.

Currently, there is [one lone mention](https://docs.python.org/3/c-api/module.html?highlight=pydoc_strvar#c.PyModuleDef) in the C-API reference, despite the fact that `PyDoc_STRVAR` is ubiquitous to `Modules/`.

Additionally, this properly uses `c:macro` within `c-api/module.html` to link.

https://bugs.python.org/issue38387

### CODE REVIEW

File: Doc/c-api/intro.rst [hunk 1]
```diff
@@ -174,6 +174,17 @@ complete listing.
    .. versionchanged:: 3.8
       MSVC support was added.
 
+.. c:macro:: PyDoc_STRVAR(name, str)
+
+   A shortcut for ``static const char name[] = str``, commonly used
```
Comment (ammaraskar) [inline Doc/c-api/intro.rst]: I don't think mentioning what this does internally is very useful, makes it somewhat contractual and could get out of date.

Maybe reword it just to be something along the lines of:

> Creates a variable with name `name` that can be used in module, function, classes etc docstrings.

File: Doc/c-api/intro.rst [hunk 2]
```diff
@@ -174,6 +174,17 @@ complete listing.
...
+   to initialize a module, function, class, or method docstring.
+
+   Example::
+
+      PyDoc_STRVAR(module_doc, "This is the module's docstring.");
```
Comment (ammaraskar) [inline Doc/c-api/intro.rst]: I think expanding this to:

```c
PyDoc_STRVAR(method_doc, "This is the method's docstring.");
// ...
    {"my_method", my_method, METH_O, method_doc},
```

would be useful to actually show a place where the macro-generated variable is used.

File: Misc/NEWS.d/next/Documentation/2019-10-06-23-44-15.bpo-38387.fZoq0S.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Document PyDoc_STRVAR macro in the C-API reference.
```
Comment (ammaraskar) [inline Misc/NEWS.d/next/Documentation/2019-10-06-23-44-15.bpo-38387.fZoq0S.rst]: Please add in a `Patch by <your preferred name>` like this https://github.com/python/cpython/blob/master/Misc/NEWS.d/next/Core%20and%20Builtins/2019-05-28-11-47-44.bpo-37077.S1h0Fc.rst

**Review by ammaraskar — CHANGES_REQUESTED** [review]

File: Misc/NEWS.d/next/Documentation/2019-10-06-23-44-15.bpo-38387.fZoq0S.rst (hunk 3 above)
Comment (bsolomon1124) [reply Misc/NEWS.d/next/Documentation/2019-10-06-23-44-15.bpo-38387.fZoq0S.rst]: Thanks for reviewing and hello from Stack Overflow! 👋

File: Doc/c-api/intro.rst [hunk 4]
```diff
@@ -174,6 +174,17 @@ complete listing.
...
+
+   .. versionchanged:: 3.8
+      This macro previously inserted ``static char name[]``.
```
Comment (ghost) [inline Doc/c-api/intro.rst]: IMO it's not worthy mentioning this `versionchanged`.

File: Misc/NEWS.d/next/Documentation/2019-10-06-23-44-15.bpo-38387.fZoq0S.rst (hunk 3 above)
Comment (aeros) [reply Misc/NEWS.d/next/Documentation/2019-10-06-23-44-15.bpo-38387.fZoq0S.rst]: @bsolomon1124 The `Patch by <name>` is optional, but you can certainly add it if you'd like to.

File: Doc/c-api/intro.rst (hunk 1 above)
Comment (aeros) [reply Doc/c-api/intro.rst]: Agreed, that's also why we typically don't list every possible use case; to avoid creating unnecessary contractual obligations. It adds to the number of potential backwards compatibility concerns.

**Comment by aeros167** [comment]
@bsolomon1124 Thanks for the PR, I think it's certainly worthwhile to document the macro, especially considering how common it's utilized.

But, I'm in agreement with the changes recommended by @ammaraskar: removing the description of the internals and adding a brief example.

**Comment by bsolomon1124** [comment]
All the suggested changes here make sense and I'll add them when I have a chance.

I also think it might be worth mentioning the stated use of these macros:

> Use the PyDoc_STR() or PyDoc_STRVAR() macro for docstrings to support building Python without docstrings (./configure --without-doc-strings).

https://www.python.org/dev/peps/pep-0007/#documentation-strings

**Comment by aeros167** [comment]
I think that might be worth adding, but I would recommend linking to the PEP itself using an inline Sphinx role:
```
Use the ``PyDoc_STR()`` or ``PyDoc_STRVAR()`` macro for docstrings to support
building Python without docstrings, as specified in :pep:`7`.
```
This provides readers a location for further information and means that we don't have to update this section if the syntax ever changes for ``./configure --without-doc-string``.

File: Doc/c-api/intro.rst (hunk 2 above)
Comment (bsolomon1124) [reply Doc/c-api/intro.rst]: I used a method docstring directly from `_collectionsmodule.c`

File: Doc/c-api/intro.rst [hunk 5]
```diff
@@ -174,6 +174,23 @@ complete listing.
    .. versionchanged:: 3.8
       MSVC support was added.
 
+.. c:macro:: PyDoc_STRVAR(name, str)
+
+   Creates a variable with name ``name`` that can be used in docstrings.
```
Comment (aeros) [inline Doc/c-api/intro.rst:179]: Formatting nit, parameter names should be *italicized* instead of in ``codeblocks``. Also, I think "called name" reads a bit better than "with name name".
```suggestion
   Creates a variable called *name* that can be used in docstrings.
```

**Review by aeros — APPROVED** [review]

File: Doc/c-api/intro.rst [hunk 6]
```diff
@@ -174,6 +174,23 @@ complete listing.
    .. versionchanged:: 3.8
       MSVC support was added.
 
+.. c:macro:: PyDoc_STRVAR(name, str)
+
+   Creates a variable called *name* that can be used in docstrings.
+
+   Use the :c:macro:`PyDoc_STRVAR` or :c:macro:`PyDoc_STR` macros for
```
Comment (ammaraskar) [inline Doc/c-api/intro.rst]: Since you mentioned `PyDoc_STR` in here, let's add documentation for it as well. Just something simple like, `Returns a docstring for the given input string`

Then the note about pep7 and example can be combined, maybe expanding the example with:
```c
...
    {"append", (PyCFunction)deque_append. METH_NOARGS, PyDoc_STR("some documentation")
...
```

**Comment by bsolomon1124** [comment]
@ammaraskar I've added a section for `PyDoc_STR` also here.  Appreciate if you have time for another review.

Screen Shot 2020-01-26 at 7 05 35 PM

File: Doc/c-api/intro.rst [hunk 7]
```diff
@@ -174,6 +174,39 @@ complete listing.
...
+
+   Creates a docstring for the given input string or an empty string
+   if docstrings are disabled.
+
+   Use :c:macro:`PyDoc_STR` in specifying docstrings to support
+   building Python without docstrings, as specified in :pep:`7`.
+
+   Example::
+
+      static PyMethodDef pysqlite_row_methods[] = {
```
Comment (aeros) [inline Doc/c-api/intro.rst:205]: Can you modify this example so that it doesn't reference any third party modules (`pysqlite` in this case)? We typically avoid doing that in the documentation unless it's specifically relevant, especially for code examples.

**Comment by aeros** [comment]
While I think it's probably okay in this case, I'd generally recommend against including additional sections in a PR that were outside of the scope of the original issue that it's attached to. The scope of this issue was specifically adding documentation for `PyDoc_STRVAR`.

Going outside of the scope can substantially to the required review time, particularly if you already have received several reviews on the PR. It may not seem like much in this case, but it can really add up when you consider the total volume of open PRs and our limited review time (it's mostly volunteer-driven). 

An exception to the above is if you specifically ask a core developer about it and they approve of including additional related changes in the PR. There's also the option of simply opening a second PR and attaching it to the same issue; assuming they're both related.

File: Doc/c-api/intro.rst (hunk 7 above)
Comment (bsolomon1124) [reply Doc/c-api/intro.rst:205]: @aeros can you clarify what you mean by 3rd-party modules?  This example is taken from CPython at https://github.com/python/cpython/blob/f669581a9527afb0d2325f9845a86715c0ba365d/Modules/_sqlite/row.c#L223.  Do you mean that we should not include any actual examples from CPython?  In that case we would want to change the `deque_methods` in this PR also.

Comment (aeros) [reply Doc/c-api/intro.rst:205]: > can you clarify what you mean by 3rd-party modules? This example is taken from CPython

Never mind, you can disregard the above comment. There's a package on PyPI titled `pysqlite` which I thought this was in reference to, I hadn't realized this was from a function in a standard library module (which is perfectly okay to use as a reference in the docs). Thanks for clearing that up.

**Review by zware — APPROVED** [review]
